### PR TITLE
[Woo POS] Improve error message when failing to check Order after a Payment Capture error

### LINKED ISSUE #13204: [Woo POS] Improve error message when failing to check Order after a Payment Capture error
If we get an error from the call to `capture_terminal_payment`, and then repeated errors from the check calls to `orders\{id}` as added in #13165, then we show a generic error message.

In this scenario, which is most likely when they have a bad network connection, we need to be clear with the merchant that the payment status is unknown.

@joe-keenan has said that we should show it as an in-line error, with an additional modal popover to prompt the merchant to check the order.

They can acknowledge (let's not use `Cancel` for the button) the error and collect payment again – it's safest to do this, as it'll benefit from the order checks by using the same order.

There are risks both ways on this – if they handle it wrong, they could double charge, or undercharge, the shopper. We need to be clear in our messaging to minimise these risks.

> We're already intercepting the error, refreshing the order, and showing "Payment successful" if find the order is actually paid. That should make it happen less often.
> But if we can't refresh the order (likely when there's a network connection issue) then we need to show an error in a way which makes it clear that they need to check whether the customer was charged or not...
> The risks we need to balance are:
> They assume the error means the customer wasn't charged, and make a new order to charge them again – but if the payment capture was successful, this means double charging the customer. We don't know either way.
> They think the error isn't relevant, e.g. because the card reader beeped acceptance of the card, and let the customer go. If the payment hadn't actually been captured, they would be undercharging the customer. (This is less likely, but could become a learnt behaviour.)
> - p1719559897287399/1718835247.409189-slack-C070SJRA8DP

### Repro steps
1. Set up Proxyman or Charles to intercept traffic for your test device
2. Launch the app using a store eligible for IPP
3. Navigate to `Menu > Point of Sale mode`
4. Add an item to your basket and tap `Checkout`
5. Enable breakpoints in your proxy for the response of `POST` requests to `https://public-api.wordpress.com/rest/v1.1/jetpack-blogs/*/rest-api/` and the requests to `GET` `https://public-api.wordpress.com/rest/v1.1/jetpack-blogs/*/rest-api/?json=true&path=/wc/v3/orders/*`
6. Tap your card on the reader
7. Abort the `capture_terminal_payment` response and `order` requests when the breakpoints are hit (equivalent to poor network conditions meaning it was lost.)
8. Observe that the payment shows a generic error – **this is what we need to improve**
10. Cancel and try again – allow the requests through
11. Observe that the "Payment success" screen is shown without needing to tap a card.

### PR DESCRIPTION
**Pull request by jaclync**
Closes: #13204 

## Why

If WooPayments captures the payment (charges the customer) but the request fails due to server issues or poor network conditions, we don't know whether the order has been paid for or not. In this case, the app first refreshes the order remotely again, and transitions to the success UI if the order is paid indeed. However, if the order refresh request fails like from poor network connections, we want to better inform the merchant about this error clearly so that the merchant does not end up undercharging or overcharging the customer.

## Design

After checking with Joe in p1719867296442819-slack-C0354HSNUJH, my understanding of the expected UX is to show an inline message like other payment errors, with two CTAs:
- Learn more: this opens a modal with more details about the error. The same modal is also shown when the message is shown
- Retry payment: this dismisses the message UI, the merchant needs to tap `Collect Payment` to trigger retry. Discussion p1719952222271239/1719867296.442819-slack-C0354HSNUJH

## How

In order to show the new inline message, a new `CardPresentPaymentEventDetails.paymentCaptureError` case was added. To emit this specific error, `CardPresentPaymentsAlertPresenterAdaptor` intercepts payment error `ServerSidePaymentCaptureError.paymentGateway(.otherError)` before the default payment error handling. The new inline message UI has view `PointOfSaleCardPresentPaymentCaptureErrorMessageView` and view model `PointOfSaleCardPresentPaymentCaptureErrorMessageViewModel`. The new modal is in `PointOfSaleCardPresentPaymentCaptureFailedView`, and is shown when the inline message view appears and when the merchant taps to learn more.

## Steps to reproduce

1. Set up Proxyman or Charles to intercept traffic for your test device
2. Launch the app using a store eligible for IPP
3. Navigate to `Menu > Point of Sale mode`
4. Add an item to your basket and tap `Checkout`
5. Enable breakpoints in your proxy for the response of `POST` requests to `https://public-api.wordpress.com/rest/v1.1/jetpack-blogs/*/rest-api/` and the requests to `GET` `https://public-api.wordpress.com/rest/v1.1/jetpack-blogs/*/rest-api/?json=true&path=/wc/v3/orders/*`
6. Tap your card on the reader
7. Abort the `capture_terminal_payment` response and `order` requests when the breakpoints are hit (equivalent to poor network conditions meaning it was lost.) --> an inline error message UI should be shown in the totals view, with a modal presented that shows more details about the error
8. Tap `I understand` --> it should dismiss the modal
9. Tap `Learn more` --> the modal should be shown again
10. Drag to dismiss the modal
11. Tap `Retry payment` and then `Collect Payment`, letting the `order` request go through when the breakpoint is hit  --> the "Payment success" screen is shown without needing to tap a card

## Screenshots

https://github.com/woocommerce/woocommerce-ios/assets/1945542/be22871e-8fda-4e03-b0b2-c91d5234c78b

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/212">19.4</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by iamgabrielma** [comment]
Code looks good, but I'm unable to reproduce the testing steps to see the modally-presented error. I might be doing something wrong with the breakpoints, as my understanding is that when this capture error happens, we have success on the server but failure on the client, so upon syncing order we should see the order created and marked as paid in the app, but on my end I don't see the order created, so it would seem that is failing the capture itself.  

For reference, these are the ones I land on:

| On failed capture attempt | On failed retry |
|--------|--------|
| Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2024-07-03 at 14 59 16 | Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2024-07-03 at 14 57 57 |

@staskus and/or @bozidarsevo can you test this one out?

**Review by staskus — COMMENTED** [review]
@iamgabrielma, I was able to reproduce the described error messages by aborting all the requests that happen after the card is tapped and payment being captured. If we abort before, we get a network error message.

https://github.com/woocommerce/woocommerce-ios/assets/4062343/e2f536be-7912-4dd7-a473-2af45957e8f9

✅ Changes work as described.

**Review by iamgabrielma — APPROVED** [review]

**Comment by iamgabrielma** [comment]
Nice, thanks for confirming and sharing the video, I was definitely cancelling too early. LGTM!

**Review by staskus — COMMENTED** [review]
One observation is that this state is impossible to quit unless the payment is retried. Should we allow returning to products view or exiting POS when issues such as these happen? I imagine such corner cases could be handled in a separate PR.

**Edit:**

I see that `Retry Payment` doesn't actually retry the payment but cancels it until `Collect Payment` is tapped again. It's a nitpick then, my only concern is that merchants could be hesitant to tap `Retry Payment` when they don't want to retry the payment and want to close the checkout instead.

image

**Review by staskus — APPROVED** [review]
✅  Looks good! 

Referencing [the discussion in this PR ](https://github.com/woocommerce/woocommerce-ios/pull/13166#discussion_r1657611687), it's still not clear if we want to create separate views for different states / error messages or find some common solution. Unless we decide to restructure the code when refining the UI, this implementation is tidy and works well 👍

**Comment by iamgabrielma** [comment]
> Should we allow returning to products view or exiting POS when issues such as these happen? I imagine such corner cases could be handled in a separate PR.

> Retry Payment doesn't actually retry the payment but cancels it until Collect Payment is tapped again. It's a nitpick then, my only concern is that merchants could be hesitant to tap Retry Payment when they don't want to retry the payment and want to close the checkout instead.

That's a good point, just for context the next milestone is all about error handling, so we'll be handling these cases more deeply soon ( p91TBi-bix-p2 )

cc: Happy to log this one if isn't already logged elsewhere @jaclync 👍

File: WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/Presented Views/PointOfSaleCardPresentPaymentCaptureFailedView.swift [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        )
+
+        static let errorDetails = NSLocalizedString(
+            "pointOfSale.cardPresentPayment.alert.paymentCaptureError.errorDetails",
+            value: "Due to an error from capturing payment and refreshing order, we couldn't load complete order information. " +
+            "To avoid undercharging or double charging, please check the latest order separately before proceeding.",
+            comment: "Subtitle of the alert presented when payment capture fails."
+        )
+
+        static let cancel = NSLocalizedString(
```
Comment (bozidarsevo) [inline WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/Presented Views/PointOfSaleCardPresentPaymentCaptureFailedView.swift]: Should we maybe rename this from `cancel` to `dismiss` or `understand` to be closer to translation or the function called when button is tapped?

Comment (jaclync) [reply WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/Presented Views/PointOfSaleCardPresentPaymentCaptureFailedView.swift]: Sure, updated in https://github.com/woocommerce/woocommerce-ios/pull/13215/commits/9934309a40e2171f5b0603aae28f3dd1d9fc2f99

**Comment by jaclync** [comment]
> One observation is that this state is impossible to quit unless the payment is retried. Should we allow returning to products view or exiting POS when issues such as these happen? I imagine such corner cases could be handled in a separate PR.
> 
> **Edit:**
> 
> I see that `Retry Payment` doesn't actually retry the payment but cancels it until `Collect Payment` is tapped again. It's a nitpick then, my only concern is that merchants could be hesitant to tap `Retry Payment` when they don't want to retry the payment and want to close the checkout instead.

Thanks @iamgabrielma and @staskus for testing and the suggestions! This is a great point where there is no way to exit POS at this point. I created an issue to include this improvement https://github.com/woocommerce/woocommerce-ios/issues/13234 along with other ones.

Merging the PR shortly!

**Pull request merged**